### DISCUSSION TITLE
JPQL이란 무엇인가요? JPA 기본 메소드와 쿼리 메소드의 차이점은?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#177] [Spring Data]
.

**Comment by nayonsoso** [comment]
### JPQL
Java Persistence Query Language

- SQL 을 추상화한 것으로, DB 벤더사에 종속적이지 않은 쿼리를 작성할 수 있게 해준다.
- 데이터베이스의 테이블과 컬럼을 참조하는 대신 엔티티의 속성을 사용해 쿼리를 작성한다.
- 이 때문에 ‘객체지향 쿼리 언어’라고 할 수 있다.
  - 이를테면, 테이블의 이름이 ‘users’ 이고 클래스 이름이 ‘User’라고 할 때, user 를 조회하는 쿼리는 다음과 같다.
  - `SELECT u FROM User u`
  - 엔티티의 속성을 참조한다는 특성 때문에 JPQL 에서 잘못된 이름의 속성을 사용하면 컴파일 에러가 발생한다. (그냥 쿼리에서 잘못된 이름의 column 을 사용하면 컴파일 에러가 뜨지 않음)

### QueryMethod
- JpaRepository 또는 CrudRepository 를 상속하는 인터페이스를 만들고, 규칙에 따라 키워드를 조합하면 QueryMethod를 만들 수 있다.
- 이름만 잘 지으면 영속 로직을 실행할 수 있기 때문에 코드가 간결해진다.
- 내부적으로는, Spring Data JPA가 QueryMethod의 이름을 분석해서 JPQL 쿼리를 생성하고, 메서드를 호출할 때 DB에 전송하는 식으로 동작한다.

### JPA의 기본 메서드
- 하지만 @Query 를 사용해 JPQL 로 만들지도 않았고, QueryMethod 를 사용하지 않았음에도 기본으로 사용할 수 있는 메서드들이 있다.
- 이는 SimpleJpaRepository에 선언되어있는 메서드들이기 때문이다.
  - Spring Data JPA는 Repository 인터페이스에 대한 구현체를 런타임에 동적으로 생성한다. 
  - 이 과정에서 Spring AOP를 활용하여, SimpleJpaRepository의 인스턴스를 프록시 객체로 사용한다.

cf. simpleJpaRepository가 제공하는 함수를 보려면 인텔리제이에서 shift 두번 눌러서 검색하면 볼 수 있다.
cf. 프록시로 simpleJpaRepository가 사용되는걸 보고 싶으면 디버깅을 해보면 된다.

**Comment by gyungchan-jo** [comment]
JPQL (Java Persistence Query Language)

JPQL은 Java Persistence API (JPA)에서 사용되는 쿼리 언어로, 객체 지향 쿼리 언어입니다. 이는 데이터베이스 테이블이 아닌 엔티티 객체를 대상으로 쿼리를 작성할 수 있게 해줍니다. SQL과 유사하지만, JPQL은 데이터베이스의 테이블과 컬럼 대신 엔티티와 그 속성을 대상으로 하고, 특정 데이터베이스에 종속되지 않으며, JPA가 SQL로 변환하여 실행합니다.

---

JPA 기본 메서드와 쿼리 메서드의 차이

JPA 기본 메소드는 JpaRepository에서 제공하는 메소드로 기본적인 CRUD 작업을 수행하며, 별도의 쿼리를 작성할 필요가 없습니다. 
- save(S entity)
- findById(ID id)
- findAll()
- deleteById(ID id)

쿼리 메소드는 메소드 이름을 기반으로 JPA가 자동으로 쿼리를 생성해주는 기능으로, 규칙에 따라 키워드를 조합하여 메소드 이름만으로 복잡한 쿼리를 작성할 수 있습니다.
- findByName(String name)
- findByNameAndAge(String name, int age)
